tests/test_validator.py

```diff
@@ -0,0 +1,94 @@
+import sys
+from typing import Any, Dict, ForwardRef, Generic, List, NamedTuple, Tuple, TypeVar, Union
+
+import pytest
+from typing_extensions import TypeAlias, TypedDict
+
+from pydantic import BaseModel, Validator
+
+ItemType = TypeVar('ItemType')
+
+NestedList = List[List[ItemType]]
+
+
+class PydanticModel(BaseModel):
+    x: int
+
+
+T = TypeVar('T')
+
+
+class GenericPydanticModel(BaseModel, Generic[T]):
+    x: NestedList[T]
+
+
+class SomeTypedDict(TypedDict):
+    x: int
+
+
+class SomeNamedTuple(NamedTuple):
+    x: int
+
+
+@pytest.mark.parametrize(
+    'tp, val, expected',
+    [
+        (PydanticModel, PydanticModel(x=1), PydanticModel(x=1)),
+        (PydanticModel, {'x': 1}, PydanticModel(x=1)),
+        (SomeTypedDict, {'x': 1}, {'x': 1}),
+        (SomeNamedTuple, SomeNamedTuple(x=1), SomeNamedTuple(x=1)),
+        (List[str], ['1', '2'], ['1', '2']),
+        (Tuple[str], ('1',), ('1',)),
+        (Tuple[str, int], ('1', 1), ('1', 1)),
+        (Tuple[str, ...], ('1',), ('1',)),
+        (Dict[str, int], {'foo': 123}, {'foo': 123}),
+        (Union[int, str], 1, 1),
+        (Union[int, str], '2', '2'),
+        (GenericPydanticModel[int], {'x': [[1]]}, GenericPydanticModel[int](x=[[1]])),
+        (GenericPydanticModel[int], {'x': [['1']]}, GenericPydanticModel[int](x=[[1]])),
+        (NestedList[int], [[1]], [[1]]),
+        (NestedList[int], [['1']], [[1]]),
```

Comment:
@dmontagu this "just worked", I was honestly kinda surprised. Should I be doing more in-depth testing here or if this works does it mean that all of our generics support works?

Reply:
I'm going to add a similar test in another branch. The following doesn't work currently (unrelated to this PR):

```python
from typing import Generic, List, TypeVar

from pydantic import BaseModel

ItemType = TypeVar('ItemType')

class GenericPydanticModel(BaseModel, Generic[ItemType]):
    x: ItemType

GenericPydanticModelList = List[GenericPydanticModel[ItemType]]
GenericPydanticModelList[int]
```

Reply:
#5183